src/seraphis_wallet/transaction_history.h

```diff
@@ -0,0 +1,195 @@
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// The goal of this component (SpTransactionHistory) is to store all the relevant information (SpTransactionStoreV1)
+// from transactions made by the wallet and allow the user to efficiently access these information later (to visualize
+// specific info from enotes or generate knowledge proofs for example).
+
+#pragma once
+
+// local headers
+#include "crypto/crypto.h"
+#include "ringct/rctTypes.h"
+#include "seraphis_core/jamtis_destination.h"
+#include "seraphis_core/jamtis_payment_proposal.h"
+#include "seraphis_impl/enote_store.h"
+#include "seraphis_main/contextual_enote_record_types.h"
+#include "seraphis_main/txtype_squashed_v1.h"
+
+// third party headers
+#include <boost/range.hpp>
+#include "boost/range/iterator_range.hpp"
+
+// standard headers
+#include <cstddef>
+#include <cstdint>
+#include <functional>
+#include <map>
+#include <tuple>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+// forward declarations
+
+using namespace sp::jamtis;
+using namespace sp;
+
+// struct to store the relevant info from a tx
+struct TransactionRecordV1
+{
+    // key images of spent enotes for tracking purposes
+    std::vector<crypto::key_image> legacy_spent_enotes;
+    std::vector<crypto::key_image> sp_spent_enotes;
+
+    // JamtisPaymentProposals
+    std::vector<JamtisPaymentProposalSelfSendV1> selfsend_payments;
+    std::vector<JamtisPaymentProposalV1> normal_payments;
+
+    // fees and total sent
+    // useful to store here also instead of looking directly at the enotes and blockchain
+    rct::xmr_amount amount_sent;
+    rct::xmr_amount fee_sent;
+};
+
+// struct to efficiently store/sort records
+struct SpTransactionStoreV1
+{
+    // quickly find TransactionRecordV1 from txid
+    serializable_unordered_map<rct::key, TransactionRecordV1> tx_records;
+
+    // sort by timestamp
+    std::vector<std::pair<uint64_t, rct::key>> txs_by_timestamp;
+};
+
+//-----------------------------------------------------------------
+/// Operators
+bool operator==(const SpTransactionStoreV1 &a, const SpTransactionStoreV1 &b);
+bool operator==(const TransactionRecordV1 &a, const TransactionRecordV1 &b);
+//-----------------------------------------------------------------
+
+class SpTransactionHistory
+{
+    SpTransactionStoreV1 m_sp_tx_store;
```

Comment:
You have now **two** classes `SpTransactionStoreV1` and `SpTransactionHistory`. What's are the reasons for two classes? Why not simply `SpTransactionHistory`?

I can't see any reasons right now, except the prediction that maybe soon a history will have transaction store V1 plus a transaction store V2 to deal with, and then you better have a class that can shield the rest of the code from the hassle that this will cause.

But well, if I am correct that fearing a transaction store V2 is unfounded, this falls away.

I have a humorous motto in the company I work: "Simple is simply simpler."